### PR TITLE
Annotate open() to enable type checking and IDE suggestions

### PR DESCRIPTION
**Pull request by GuyAglionby**
First, thanks for the package! Super useful.

This change moves reading and writing method definitions to `ReaderWriterBase` so IDEs can suggest methods when using the `jsonlines.open` convenience syntax. I follow `builtins.open` in raising an `io.UnsupportedOperation` exception when bad methods are called.

This may make the interface for `Reader` and `Writer` a little unintuitive, but this seems the cleanest way to do it. An alternative would be to keep `ReaderWriterBase` as is, and wrap responses from `.open` in a new parent class. Please let me know your thoughts - I'm happy to iterate, and just provide this to give an idea of what it'd look like.

### CODE REVIEW

**Comment by wbolster** [comment]
hmmm... i'm not entirely sure i understand what the actual problem is. `Reader` has read methods, `Writer` has write methods, so far so good right? but the return type from `open()` is inferred by whatever tool it is you're using?

perhaps adopting type annotations could help here? :thinking:

**Comment by GuyAglionby** [comment]
That's it - PyCharm 2020.3.2 isn't able to infer whether the `Reader` or `Writer` is returned (I guess it doesn't check the [conditional](https://github.com/wbolster/jsonlines/blob/master/jsonlines/jsonlines.py#L323-L326)), and just gives `Union[Reader, Writer]`. I assume the methods suggested in autocomplete are then only those in the common superclass. Moving the methods there is thus how this change makes the suggestions appear, rather than because the inferred type has changed.

Method [overloading](https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading) looks like the right sort of thing but only checks for arguments of different types, not different values. I'm not sure what other hinting exists - I couldn't find anything after a look around.

I realise this isn't major, but would be a nice quality of life improvement!

**Comment by wbolster** [comment]
`typing.overload` can be used with `typing.Literal`. see [PEP 586 rationale](https://www.python.org/dev/peps/pep-0586/#motivation-and-rationale) for an example that is pretty much the same as the ‘problem’ with `jsonlines.open()`:

> `open(filename, mode)` returns either `IO[bytes]` or `IO[Text]` depending on whether the second argument is something like `r` or `rb`.

**Comment by GuyAglionby** [comment]
Perfect, thanks for the pointer. I've implemented this and it works as expected in PyCharm. It's only available in 3.8+ so I've included an import exception check. Please let me know any thoughts.

**Comment by GuyAglionby** [comment]
Hi @wbolster, when you have time I think this is complete

**Comment by wbolster** [comment]
oh @GuyAglionby sorry i didn't realise you actually tried out the type hinting approach. so that works with the tools you use? (pycharm?) or are there imperfections or other trade-offs?

**Comment by GuyAglionby** [comment]
No problem :) Yep, looks like it works.

Screen Shot 2021-02-01 at 15 44 15
Screen Shot 2021-02-01 at 15 44 46
Screen Shot 2021-02-01 at 15 44 24

I think it's as simple as that. Only downside is that `Literal` is only around in 3.8+, so the fix doesn't work in lower versions but nothing breaks.

**Comment by wbolster** [comment]
cool, thanks, let's take this approach in that case. type annotations are more in line with the direction the ecosystem is taking anyway (see e.g. #59).

**Comment by wbolster** [comment]
hey @GuyAglionby i took the liberty to push to this pull request to address a few issues:

- typing_extensions dependency (only for python < 3.8) to ensure `Literal` is available
- unconditional overload definitions, which fixes mypy warnings (see also #59)
- add more annotations (see #59)
- change first arg name to jsonlines.open() to match builtins.open()

can you confirm this still works for you in its current shape?

**Review by wbolster — APPROVED** [review]

File: jsonlines/jsonlines.py [hunk 1]
```diff
@@ -295,27 +305,17 @@ def write_all(self, iterable):
...
 
 
-try:
-    from typing import Literal, overload
-
-    @overload
-    def open(name, **kwargs) -> Reader:
-        ...
+@overload
+def open(file: _Openable, mode: Literal["r"], **kwargs) -> Reader:
```
Comment (GuyAglionby) [inline jsonlines/jsonlines.py:309]: If we change to `mode: Literal["r"] = "r"` then the type annotations will appear when the arg isn't specified.

Comment (wbolster) [reply jsonlines/jsonlines.py]: thanks, addressed in 05b810f67eaeae25a48089a204aa8b46711e820e

**Comment by wbolster** [comment]
merged 🚀 thanks again!

**Comment by GuyAglionby** [comment]
Cheers!

**Comment by wbolster** [comment]
this effort continues in #62/#59

**Pull request merged**